src/v/transform/transform_processor.cc

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "transform/transform_processor.h"
+
+#include "model/record_batch_reader.h"
+#include "model/timeout_clock.h"
+#include "random/simple_time_jitter.h"
+#include "transform/logger.h"
+#include "wasm/api.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/queue.hh>
+#include <seastar/core/sleep.hh>
+
+namespace transform {
+
+namespace {
+model::record_batch_reader::data_t
+extract_batches(model::record_batch_reader::storage_t s) {
+    if (std::holds_alternative<model::record_batch_reader::data_t>(s)) {
+        return std::get<model::record_batch_reader::data_t>(std::move(s));
+    } else {
+        const auto& f = std::get<model::record_batch_reader::foreign_data_t>(s);
+        model::record_batch_reader::data_t copied;
+        for (const model::record_batch& batch : *f.buffer) {
+            copied.push_back(batch.copy());
+        }
+        return copied;
+    }
+}
+
+ss::future<std::optional<model::offset>> consume_batches(
```

Comment:
i guess i'm wondering why this isn't a normal consumer passed to a `record_batch_reader::consume` ?

Reply:
The consume interface works at the individual batch level, and I wanted to always make sure that we were pipelining reads in another implementation, anyways I will simplify this now, until the optimized version. Apologies for the noise.